src/internal/collection/postgres_listener.go

```diff
@@ -169,62 +174,75 @@ func parsePostgresEpoch(s string) (time.Time, error) {
-	if len(parts) != 6 {
+	parts := strings.Split(payload, " ")
+	if len(parts) != 7 {
 		return &postgresEvent{err: errors.Errorf("failed to parse notification payload, wrong number of parts: %d", len(parts))}
 	}
 
-	value, err := base64.StdEncoding.DecodeString(parts[1])
+	value, err := base64.StdEncoding.DecodeString(parts[4])
 	if err != nil {
 		return &postgresEvent{err: errors.Wrap(err, "failed to decode notification payload index value base64")}
 	}
 
-	key, err := base64.StdEncoding.DecodeString(parts[2])
+	key, err := base64.StdEncoding.DecodeString(parts[0])
 	if err != nil {
 		return &postgresEvent{err: errors.Wrap(err, "failed to decode notification payload key base64")}
 	}
 
-	eventType := watch.EventError
-	switch parts[3] {
+	result := &postgresEvent{
+		index:     parts[3],
+		value:     string(value),
+		key:       string(key),
+		eventType: watch.EventError,
+	}
+
+	switch parts[2] {
 	case "INSERT", "UPDATE":
-		eventType = watch.EventPut
+		result.eventType = watch.EventPut
 	case "DELETE":
-		eventType = watch.EventDelete
+		result.eventType = watch.EventDelete
 	}
-	if eventType == watch.EventError {
+	if result.eventType == watch.EventError {
 		return &postgresEvent{err: errors.Errorf("failed to decode notification payload operation type: %s", parts[3])}
 	}
 
-	eventTime, err := parsePostgresEpoch(parts[4])
+	result.time, err = parsePostgresEpoch(parts[1])
 	if err != nil {
 		return &postgresEvent{err: errors.Wrapf(err, "failed to decode notification payload timestamp: %s", parts[4])}
 	}
 
-	protoData, err := base64.StdEncoding.DecodeString(parts[5])
-	if err != nil {
-		return &postgresEvent{err: errors.Wrap(err, "failed to decode notification payload proto base64")}
+	switch parts[5] {
+	case "inline":
+		result.protoData, err = base64.StdEncoding.DecodeString(parts[6])
+		if err != nil {
+			return &postgresEvent{err: errors.Wrap(err, "failed to decode notification payload proto base64")}
+		}
+	case "stored":
+		result.storedID = parts[6]
 	}
 
-	return &postgresEvent{
-		index:     parts[0],
-		value:     string(value),
-		key:       string(key),
-		eventType: eventType,
-		time:      eventTime,
-		protoData: protoData,
-	}
+	return result
 }
 
-func (pe *postgresEvent) WatchEvent(template proto.Message) *watch.Event {
+func (pe *postgresEvent) WatchEvent(ctx context.Context, db *sqlx.DB, template proto.Message) *watch.Event {
 	if pe.err != nil {
 		return &watch.Event{Err: pe.err, Type: watch.EventError}
 	}
 	if pe.eventType == watch.EventDelete {
 		// Etcd doesn't return deleted row values - we could, but let's maintain parity
 		return &watch.Event{Key: []byte(pe.key), Type: pe.eventType, Template: template}
 	}
+	if pe.protoData == nil && pe.storedID != "" {
+		// The proto data was too large to fit in the payload, read it from a temporary location.
+		if err := db.QueryRowContext(ctx, "select proto from collections.large_notifications where id = $1", pe.storedID).Scan(&pe.protoData); err != nil {
```

Comment:
Here's where we read the proto data from the `large_notifications` table - when we're delivering the event to the channel the user is reading from.  We could possibly make this more efficient in the future by caching these in case multiple listeners are receiving the same event, but it isn't worth the complexity at this point.